warp10/src/main/java/io/warp10/script/functions/TOGTS.java

```diff
@@ -16,98 +16,257 @@
 
 package io.warp10.script.functions;
 
-import java.math.BigDecimal;
-import java.nio.charset.StandardCharsets;
-import java.util.HashMap;
-import java.util.Map;
-
-import org.apache.thrift.TDeserializer;
-import org.apache.thrift.TException;
-import org.apache.thrift.protocol.TCompactProtocol;
-
 import io.warp10.continuum.gts.GTSDecoder;
 import io.warp10.continuum.gts.GTSEncoder;
 import io.warp10.continuum.gts.GTSHelper;
 import io.warp10.continuum.gts.GTSWrapperHelper;
 import io.warp10.continuum.gts.GeoTimeSerie;
+import io.warp10.continuum.gts.MetadataSelectorMatcher;
 import io.warp10.continuum.store.thrift.data.GTSWrapper;
 import io.warp10.crypto.OrderPreservingBase64;
 import io.warp10.script.NamedWarpScriptFunction;
 import io.warp10.script.WarpScriptException;
 import io.warp10.script.WarpScriptStack;
 import io.warp10.script.WarpScriptStackFunction;
+import org.apache.thrift.TDeserializer;
+import org.apache.thrift.TException;
+import org.apache.thrift.protocol.TCompactProtocol;
+
+import java.math.BigDecimal;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
 
 /**
  * Converts an encoder into a map of gts, one per type
+ * TODO: Provide a map type->selector or type->list of selectors, with an optional 'label' key. By default, the enforced type will be added in a ".type" label.
+ * label = null will not add any label to the output GTS.
  */
 public class TOGTS extends NamedWarpScriptFunction implements WarpScriptStackFunction {
-  
+
+  private static final String LABEL_NAME_PARAMETER = "label";
+  private static final String DEFAULT_LABEL_NAME = ".type";
+
   public TOGTS(String name) {
     super(name);
-  }  
+  }
 
   @Override
   public Object apply(WarpScriptStack stack) throws WarpScriptException {
+
+    Map<String, ArrayList<MetadataSelectorMatcher>> typeMap = null;
     Object top = stack.pop();
-    
-    if (!(top instanceof String) && !(top instanceof byte[]) && !(top instanceof GTSEncoder)) {
-      throw new WarpScriptException(getName() + " operates on a string, byte array or encoder.");
+    // this is the default label name.
+    String extraLabel = DEFAULT_LABEL_NAME;
+
+    if (top instanceof Map) {
+      typeMap = new LinkedHashMap<String, ArrayList<MetadataSelectorMatcher>>();
```

Comment:
Why use a `LinkedHashMap` even though the given Map does not guarantee order?

Reply:
input *is* a LinkedHashMap